### PR TITLE
plat/kvm/arm64: Firecracker support on arm64 

### PR DESCRIPTION
**Pull request by michpappas**
### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.

### Base target

 - Architecture(s): [`arm_64`]
 - Platform(s): [`kvm`]
 - Application(s): [N/A]

### Additional configuration

Enables using `CONFIG_KVM_VMM_FIRECRACKER` on `arm64`.

### Description of changes

This PR enables Firecracker support on `arm64`. It introduces a set of boot pagetables for Firecracker and updates `Config.uk` to allow using `CONFIG_KVM_VMM_FIRECRACER` on `arm64`. Remaining work such as the linux boot probotol on arm64 and ns16550 are handled on depending PRs.

GitHub-Depends-On: #988
GitHub-Depends-On: #986 
GitHub-Depends-On: #985
GitHub-Depends-On: #912

### CODE REVIEW

File: plat/kvm/arm/firecracker_bpt64.S [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ */
+
+#include <arm/arm64/pagetable.S>
+
+#include <arch/reloc.h>
+#include <uk/config.h>
+
+/* ------------------- Memory Map of Firecracker on Arm64 -----------------
+ *
+ * 0x0000000000000000 - 0x0000000007ffffff	Devices:     0 -    2GiB
```
Comment (zhxj9823) [inline plat/kvm/arm/firecracker_bpt64.S]: 2GiB should correspond to `0x000000007fffffff` instead of `0x0000000007fffff`.

Comment (michpappas) [reply plat/kvm/arm/firecracker_bpt64.S]: Thanks, fixed

File: plat/kvm/include/kvm-arm64/image.h [hunk 2]
```diff
@@ -6,7 +6,11 @@
 #ifndef __KVM_ARM64_IMAGE_H__
 #define __KVM_ARM64_IMAGE_H__
 
+#if defined(CONFIG_KVM_VMM_QEMU)
```
Comment (mogasergiu) [inline plat/kvm/include/kvm-arm64/image.h]: Now that Simon mentioned it. Maybe we can just go with `#if`?

File: plat/kvm/arm/qemu_bpt64.S [hunk 3]
```diff
@@ -0,0 +1,117 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <arm/arm64/pagetable.S>
+
+#include <arch/reloc.h>
```
Comment (mogasergiu) [inline plat/kvm/arm/qemu_bpt64.S]: I believe it should be `uk/reloc.h` now

File: plat/kvm/arm/firecracker_bpt64.S [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+.align 12
+arm64_bpt_l3_pt0:
+	ur_pte  arm64_bpt_l2_pt0, PTE_TYPE_TABLE
+	ur_pte  arm64_bpt_l2_pt1, PTE_TYPE_TABLE
+	pte_zero	, 509
+#ifdef CONFIG_PAGING
+	ur_pte  arm64_bpt_l2_pt511, PTE_TYPE_TABLE
+#else
+	pte_zero	, 1
+#endif /* CONFIG_PAGING */
```
Comment (mogasergiu) [inline plat/kvm/arm/firecracker_bpt64.S]: `!CONFIG_PAGING`

File: plat/kvm/arm/firecracker_bpt64.S [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+ * 0x0000008000000000 - 0x000000ffffffffff	Table descriptor to l2_pt1
+ * 0x0000010000000000 - 0x0000ff7fffffffff	Unmapped
+ * 0x0000ff8000000000 - 0x0000ffffffffffff	Table descriptor to l2_pt511
+ */
+.align 12
+arm64_bpt_l3_pt0:
+	ur_pte  arm64_bpt_l2_pt0, PTE_TYPE_TABLE
+	ur_pte  arm64_bpt_l2_pt1, PTE_TYPE_TABLE
+	pte_zero	, 509
+#ifdef CONFIG_PAGING
```
Comment (mogasergiu) [inline plat/kvm/arm/firecracker_bpt64.S]: Ditto for `#if`

File: plat/kvm/arm/firecracker_bpt64.S [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+/* L2: 512GiB - 1TiB (1GiB / entry)
+ *
+ * 0x0000008000000000 - 0x000000ffffffffff	RAM @ 512GiB
+ */
+.align 12
+arm64_bpt_l2_pt1:
+	pte_fill	0x0000008000000000, 512, 2, PTE_BLOCK_NORMAL_RW
+
+#ifdef CONFIG_PAGING
```
Comment (mogasergiu) [inline plat/kvm/arm/firecracker_bpt64.S]: Ditto :bird:

File: plat/kvm/arm/qemu_bpt64.S [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+.align 12
+arm64_bpt_l3_pt0:
+	ur_pte  arm64_bpt_l2_pt0, PTE_TYPE_TABLE
+	ur_pte  arm64_bpt_l2_pt1, PTE_TYPE_TABLE
+	pte_zero	, 509
+#ifdef CONFIG_PAGING
+	ur_pte  arm64_bpt_l2_pt511, PTE_TYPE_TABLE
+#else
+	pte_zero	, 1
+#endif /* CONFIG_PAGING */
```
Comment (mogasergiu) [inline plat/kvm/arm/qemu_bpt64.S]: `!CONFIG_PAGING`

File: plat/kvm/arm/qemu_bpt64.S [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+ * 0x0000ff8000000000 - 0x0000ffffffffffff	Table descriptor to l2_pt511
+ */
+.align 12
+arm64_bpt_l3_pt0:
+	ur_pte  arm64_bpt_l2_pt0, PTE_TYPE_TABLE
+	ur_pte  arm64_bpt_l2_pt1, PTE_TYPE_TABLE
+	pte_zero	, 509
+#ifdef CONFIG_PAGING
+	ur_pte  arm64_bpt_l2_pt511, PTE_TYPE_TABLE
+#else
```
Comment (mogasergiu) [inline plat/kvm/arm/qemu_bpt64.S]: `/* CONFIG_PAGING */` :dog:

File: plat/kvm/arm/firecracker_bpt64.S [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+ * 0x0000ff8000000000 - 0x0000ffffffffffff	Table descriptor to l2_pt511
+ */
+.align 12
+arm64_bpt_l3_pt0:
+	ur_pte  arm64_bpt_l2_pt0, PTE_TYPE_TABLE
+	ur_pte  arm64_bpt_l2_pt1, PTE_TYPE_TABLE
+	pte_zero	, 509
+#ifdef CONFIG_PAGING
+	ur_pte  arm64_bpt_l2_pt511, PTE_TYPE_TABLE
+#else
```
Comment (mogasergiu) [inline plat/kvm/arm/firecracker_bpt64.S]: `/* CONFIG_PAGING */` :dodo:

File: plat/kvm/arm/qemu_bpt64.S [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+/* L2: 512GiB - 1TiB (1GiB / entry)
+ *
+ * 0x0000008000000000 - 0x000000ffffffffff	PCIe hi-mem @ 512GiB
+ */
+.align 12
+arm64_bpt_l2_pt1:
+	pte_fill	0x0000008000000000, 512, 2, PTE_BLOCK_DEVICE_nGnRE
+
+#ifdef CONFIG_PAGING
```
Comment (mogasergiu) [inline plat/kvm/arm/qemu_bpt64.S]: Ditto for `#if`

File: plat/kvm/arm/qemu_bpt64.S [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+ * 0x0000008000000000 - 0x000000ffffffffff	Table descriptor to l2_pt1
+ * 0x0000010000000000 - 0x0000ff7fffffffff	Unmapped
+ * 0x0000ff8000000000 - 0x0000ffffffffffff	Table descriptor to l2_pt511
+ */
+.align 12
+arm64_bpt_l3_pt0:
+	ur_pte  arm64_bpt_l2_pt0, PTE_TYPE_TABLE
+	ur_pte  arm64_bpt_l2_pt1, PTE_TYPE_TABLE
+	pte_zero	, 509
+#ifdef CONFIG_PAGING
```
Comment (mogasergiu) [inline plat/kvm/arm/qemu_bpt64.S]: Ditto :parrot:

**Review by mogasergiu — CHANGES_REQUESTED** [review]
Minor repetitive comments for you to lose your sanity :service_dog:.

File: plat/kvm/include/kvm-arm64/image.h (hunk 2 above)
Comment (michpappas) [reply plat/kvm/include/kvm-arm64/image.h]: Yes will fix thanks

File: plat/kvm/arm/firecracker_bpt64.S (hunk 9 above)
Comment (mogasergiu) [reply plat/kvm/arm/firecracker_bpt64.S]: Sorry meant with an `!` of course :nerd_face: :point_up: 

EDIT: No, I am dumb it's fine :man_facepalming:  :laughing:

**Review by mogasergiu — APPROVED** [review]
FC, meet the superior architecture! :muscle:

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request closed without merging**